addons/survey/controllers/main.py

```diff
@@ -490,161 +490,69 @@ def _generate_report(self, user_input, download=True):
-        scoring_failed_count = 0
-        for count_data_item in count_data:
-            if count_data_item['scoring_success']:
-                scoring_success_count = count_data_item['scoring_success_count']
+    def _extract_filters_data(self, survey, post):
+        search_filters = []
+        line_filter_domain, line_choices = [], []
+        for data in post.get('filters', '').split('|'):
+            try:
+                row_id, answer_id = data.split(',')
+                row_id = int(row_id)
+                answer_id = int(answer_id)
+            except:
+                pass
             else:
-                scoring_failed_count = count_data_item['scoring_success_count']
-
-        graph_data = [{
-            'text': _('Passed'),
-            'count': scoring_success_count,
-            'color': '#2E7D32'
-        }, {
-            'text': _('Missed'),
-            'count': scoring_failed_count,
-            'color': '#C62828'
-        }]
-
-        total_scoring_success = scoring_success_count + scoring_failed_count
-        return {
-            'success_rate': round((scoring_success_count / total_scoring_success) * 100, 1) if total_scoring_success > 0 else 0,
-            'graph_data': graph_data
-        }
+                if row_id and answer_id:
+                    new_domain = ['&', ('matrix_row_id', '=', row_id), ('suggested_answer_id', '=', answer_id)]
+                    line_filter_domain = expression.AND([new_domain, line_filter_domain])
+                    answers = request.env['survey.question.answer'].browse([row_id, answer_id])
+                elif answer_id:
+                    line_choices.append(answer_id)
+                    answers = request.env['survey.question.answer'].browse([answer_id])
+                if answer_id:
+                    search_filters.append({
+                        'question': answers[0].question_id.title,
+                        'answers': '%s%s' % (answers[0].value, ': %s' % answers[1].value if len(answers) > 1 else '')
+                    })
+        if line_choices:
+            line_filter_domain = expression.AND([[('suggested_answer_id', 'in', line_choices)], line_filter_domain])
+
+        if line_filter_domain:
+            # here we go from input lines to user input to avoid a domain like
+            # ('user_input.survey_id', '=', survey.id) which would return a lot of data
+            # on huge surveys, especially with test entries and draft entries in mind
+            matching_user_inputs = request.env['survey.user_input.line'].sudo().search(line_filter_domain).mapped('user_input_id')
+            if post.get('finished'):
+                user_input_lines = matching_user_inputs.filtered(lambda ui: not ui.test_entry and ui.state == 'done').mapped('user_input_line_ids')
+            else:
+                user_input_lines = matching_user_inputs.filtered(lambda ui: not ui.test_entry and ui.state != 'new').mapped('user_input_line_ids')
```

Comment:
That code is a bit messy indeed. I vote to keep it as it is currently. I bet we will improve filtering on results soon and it will probably be an opportunity to clean that brol (TM).

Reply:
Ok, ok. I don't want you to have a bad night.